ospd_openvas/preferencehandler.py

```diff
@@ -0,0 +1,610 @@
+import binascii
+
+from enum import IntEnum
+from typing import Optional, Dict, List, Tuple
+from base64 import b64decode
+
+from ospd.scan import ScanCollection
+from ospd_openvas.openvas import Openvas
+from ospd_openvas.db import KbDB
+
+logger = logging.getLogger(__name__)
+
+
+OID_SSH_AUTH = "1.3.6.1.4.1.25623.1.0.103591"
+OID_SMB_AUTH = "1.3.6.1.4.1.25623.1.0.90023"
+OID_ESXI_AUTH = "1.3.6.1.4.1.25623.1.0.105058"
+OID_SNMP_AUTH = "1.3.6.1.4.1.25623.1.0.105076"
+OID_PING_HOST = "1.3.6.1.4.1.25623.1.0.100315"
+
+
+class AliveTest(IntEnum):
+    """ Alive Tests. """
+
+    ALIVE_TEST_TCP_ACK_SERVICE = 1
+    ALIVE_TEST_ICMP = 2
+    ALIVE_TEST_ARP = 4
+    ALIVE_TEST_CONSIDER_ALIVE = 8
+    ALIVE_TEST_TCP_SYN_SERVICE = 16
+
+
+def _from_bool_to_str(value: int) -> str:
+    """ The OpenVAS scanner use yes and no as boolean values, whereas ospd
+    uses 1 and 0."""
+    return 'yes' if value == 1 else 'no'
+
+
+class PreferenceHandler:
+    def __init__(
+        self, scan_id: str, kbdb: KbDB, scan_collection: ScanCollection
+    ):
+        self.scan_id = scan_id
+        self.kbdb = kbdb
+        self.scan_collection = scan_collection
+
+        self._openvas_scan_id = None
+
+        self._target_options = None
+
+    @property
+    def openvas_scan_id(self) -> str:
+        """ Create the openvas scan id and store it in the redis kb.
+        Return the openvas scan_id.
+        """
+        if self._openvas_scan_id is not None:
+            return self._openvas_scan_id
+
+        self._openvas_scan_id = str(uuid.uuid4())
+        self.kbdb.add_scan_id(self.scan_id, self._openvas_scan_id)
+
+        return self._openvas_scan_id
+
+    @property
+    def target_options(self) -> Dict:
+        """ Return target options from Scan collection """
+        if self._target_options is not None:
+            return self._target_options
+
+        self._target_options = self.scan_collection.get_target_options(
+            self.scan_id
+        )
+        return self._target_options
+
+    def get_vts_in_groups(
+        self, filters: List[str], vts_cache: Dict[str, Dict],
+    ) -> List[str]:
+        """ Return a list of vts which match with the given filter.
+
+        Arguments:
+            filters A list of filters. Each filter has key, operator and
+                    a value. They are separated by a space.
+                    Supported keys: family
+
+        Returns a list of vt oids which match with the given filter.
+        """
+        vts_list = list()
+        families = dict()
+
+        for oid in vts_cache:
+            family = vts_cache[oid]['custom'].get('family')
+            if family not in families:
+                families[family] = list()
+
+            families[family].append(oid)
+
+        for elem in filters:
+            key, value = elem.split('=')
+            if key == 'family' and value in families:
+                vts_list.extend(families[value])
+
+        return vts_list
+
+    def get_vt_param_type(
```

Comment:
Oh btw. are the get_vt(s)_ methods used outside of this class? if not they should be marked as internal by using an underscore prefix in their name `def _get_vt_`